docs/okeps/okep-5377-extend-udn-to-support-multiple-cluster-subnets-in-layer3-topology.md

```diff
@@ -0,0 +1,284 @@
+subnets and enabling controlled mutability.
+
+
+## User-Stories/Use-Cases
+
+### Story 1: Expanding subnet pool for growing clusters
+
+**As a** cluster operator,
+**I want** to expand the available IP address pool for a Layer3 network,
+**so that** I can scale the cluster without redeploying or disrupting workloads.
+
+
+## Proposed Solution
+
+### API Details
+
+1. Update `UDN`/`CUDN` CRD to allow `layer3.subnets` to accept multiple subnets per IP family.
+2. Allow updates to the `layer3.subnets` field, but only to **append** new entries.
+   - Reject updates that attempt to remove or modify existing subnet entries.
+   - Reject new entries that overlap with existing subnets or use a different `hostSubnet`.
+
+
+### Implementation Details
+
+### Node Subnet Allocation
+
+The current NAD implementation already supports multiple subnets in Layer3 topology. However,
+when a new subnet is added to a NAD, the corresponding network controller is first deleted and then
+recreated with the updated subnet list. This behavior can cause node subnets to be reallocated,
+resulting in unnecessary disruption.
+
+In practice, this issue can often be worked around by restarting the `ovnkube-cluster-manager`, so
+that the controller is recreated with the latest NAD spec and reserve node subnets based on the
+`k8s.ovn.org/node-subnets` annotation.
+
+Ideally, nodes should retain their originally assigned subnets when adding new cluster subnets.
+Changes are required in the `SecondaryNetworkClusterManager` to ensure stable node subnet assignments
+when new subnets are dynamically added.
+
+### OVN Network Topology
+
+Adding new subnets to a Layer3 network simply extends its cluster subnet allocation pool. Each node
+continues to allocate one subnet from the available pool. The overall OVN network topology remains
+largely unchanged. However, one additional change is required in the routing policies of the OVN
+Cluster Router: the `DefaultNoReroutePriority` (priority 102) policies.
+
+Currently it only allows east-west traffic within the individual cluster subnet.
+
+For example, for a network with cluster subnets `10.1.0.0/16` and `10.2.0.0/16`, the OVN cluster
+router policy list includes:
+```
+$ ovn-nbctl lr-policy-list  udn_udn.primary.layer3_ovn_cluster_router
+Routing Policies
+      ...
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+```
+
+To enable east-west traffic between the subnets, the following additional policies must be added:
+```
+$ ovn-nbctl lr-policy-list  udn_udn.primary.layer3_ovn_cluster_router
+Routing Policies
+      ...
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+      102   ip4.src == 10.1.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.2.0.0/16    allow
+      102   ip4.src == 10.2.0.0/16 && ip4.dst == 10.1.0.0/16    allow
+```
+
+
+### API Validations
```

Comment:
What about API validation within OVNK? For example, lets say i add another subnet, and it collides with the join subnet. Normally OVNK would refuse to render this NAD, and put the UDN in error state. Can you document what happens in this case? Does the NAD get left as is, and the UDN now goes to error state?